### ISSUE TITLE
Saroo as a RAM cartridge with Fenrir.

### ISSUE THREAD

**Issue by atds55** [tpunix/SAROO#224]
Hello.
I tested the Saroo operating as a RAM cartridge with the Fenrir, and it did not work correctly. Some games did not detect the Saroo as a RAM cartridge, and the games that detected the Saroo as a RAM cartridge had corrupted graphics, and after a few minutes they went to a black screen or went to the CD player screen.
No game that uses a 1MB RAM cartridge detected the Saroo as a RAM cartridge.
I just wanted to leave my report, because I read on the Chinese forum that you were unable to test this function with the Fenrir, you only tested it with the Rhea/Phoebe.

I have a Saturn VA 0.8 and an old Fenrir 20pin. I tested with Saroo on FW 0.7.

https://github.com/user-attachments/assets/6cf4f422-277c-451f-8416-d1f812e1519f

https://github.com/user-attachments/assets/008617da-5fe4-4482-b7be-ceb49bcf1489

https://github.com/user-attachments/assets/43e3cd1c-5f11-4f48-8bd5-d9ca1b9bc0e7

https://github.com/user-attachments/assets/4ff69157-9c4b-499f-9e17-d489a44a0a7e

https://github.com/user-attachments/assets/c47ec4a9-b1b2-4a32-ae36-4572e82b1699

I decided to post this information here, as I don't have an account on the Chinese forum, and I wasn't able to register.

**Comment by ragingchaos** [comment]
This wasn't a problem in 0.6, you just had to load the game with the X button loader in Fenrir's menu to get it to work with Saroo's RAM. If you use the default A button you will always get corruption.

Firm 0.7 has broken this, the X button loader now causes Fenrir to either crash or boot loop back into Saroo.

**Comment by whitezombie2000** [comment]
I got a Saroo about a month ago. I tested it with a Japanese Model 2 with a Phoebe, Japanese Model 2 with a CD-ROM, an American Model 1 MK-80000 with a Rhea, and an American Model 1 MK-80000 with a 20 pin Fenrir. I think I did all my testing with 0.6 using Street Fighter Zero 3 to test the RAM and had no issue. I tested the memory cart function with Dragon Force and Albert Odyssey with no issue. I did upgrade the cart to v.07 so might need to re-test this. 

One problem I have noticed with all the consoles using Saroo is an annoy popping sound at the beginning of music tracks that someone reported here. I hear it frequently in Albert Odyssey.
https://github.com/tpunix/SAROO/issues/199

**Comment by yzbss** [comment]
Do not enter the system player screen, select the third item in the saroo screen directly to start the disc game
Try adding the following parameter under global
[global]
lang_id = 1
sort_mode = 1
M_201FF80=D101412B
M_201FF84=E4000009
M_201FF88=02006364
M_6000288=0201FF80

**Comment by whitezombie2000** [comment]
> This wasn't a problem in 0.6, you just had to load the game with the X button loader in Fenrir's menu to get it to work with Saroo's RAM. If you use the default A button you will always get corruption.
> 
> Firm 0.7 has broken this, the X button loader now causes Fenrir to either crash or boot loop back into Saroo.

I can confirm this boot loop. I get the same behavior with v0.7 and pressing X using a 20 pin Fenrir. I put the v.06 files back on the SD card, it works fine pressing X, games with RAM load.

**Comment by atds55** [comment]
> > This wasn't a problem in 0.6, you just had to load the game with the X button loader in Fenrir's menu to get it to work with Saroo's RAM. If you use the default A button you will always get corruption.
> > Firm 0.7 has broken this, the X button loader now causes Fenrir to either crash or boot loop back into Saroo.
> 
> I can confirm this boot loop. I get the same behavior with v0.7 and pressing X using a 20 pin Fenrir. I put the v.06 files back on the SD card, it works fine pressing X, games load fine, and games with RAM load.
> 
> A side note that the Saroo, Fenrir, and Fenrir Loader Kai menu don't seem to play nice together so use the Fenrir menu only.

Yes, using this method Saroo works fine as a 1MB/4MB cartridge and KOF95 ROM cartridge in fw V0.6.

> Não entre na tela do player do sistema, selecione o terceiro item diretamente na tela do saroo para iniciar o jogo do disco Tente adicionar o seguinte parâmetro em global (apenas para v0.7) [global] lang_id = 0 sort_mode = 1 M_201FF80=D101412B M_201FF84=E4000009 M_201FF88=02006364 M_6000288=0201FF80

This method didn't work here, and now when I load the game through the Fenrir menu the Sega logo doesn't appear.

**Comment by ragingchaos** [comment]
> Fenrir Loader Kai menu don't seem to play nice together

A temp work around is to try and load FLK cheats with the X button and when the cheat menu pops up press ABC+Start to return to the CD player menu. Clicking "Start Application" from there loads the game correctly.

**Comment by tzmwx** [comment]
> Hello. I tested the Saroo operating as a RAM cartridge with the Fenrir, and it did not work correctly. Some games did not detect the Saroo as a RAM cartridge, and the games that detected the Saroo as a RAM cartridge had corrupted graphics, and after a few minutes they went to a black screen or went to the CD player screen. No game that uses a 1MB RAM cartridge detected the Saroo as a RAM cartridge. I just wanted to leave my report, because I read on the Chinese forum that you were unable to test this function with the Fenrir, you only tested it with the Rhea/Phoebe.
> 
> I have a Saturn VA 0.8 and an old Fenrir 20pin. I tested with Saroo on FW 0.7.
> 
>  Marvel.vs.Street.Fighter.-.Fenrir.+.Saroo.mp4 
> 
>  X-Men.vs.Street.Fighter.-.Fenrir.+.Saroo.mp4 
> 
>  TKOF.95.4MB.RAM.-.Fenrir.+.Saroo.mp4 
> 
>  dungeons.and.dragons.shadow.over.mystara.-.Fenrir.+.Saroo.mp4 
> 
>  SF.Zero.3.-.Fenrir.+.Saroo.mp4 
> 
> I decided to post this information here, as I don't have an account on the Chinese forum, and I wasn't able to register.

To use fenrir, you need to press the X key, but this mode is unstable and may not guarantee the accuracy of the data.

Test method, remove the saroo card and use a multi-track game mirror image（like KOF95、SegaRally……） to test.PRESS the X key（Not A key）, Can your fenrir recognize all tracks every time?

Can you provide some testing information for this malfunction? Pay attention to 'removing saroo'

For example, the game KOF95 should have 1 data track and 26 music tracks, totaling 27 tracks. However, in X mode, FENRIR sometimes has 27 tracks and sometimes has 2 tracks. The accurate number is 27


https://github.com/user-attachments/assets/ac6e69d0-2336-4b86-876f-ea1e7607dbae


https://github.com/user-attachments/assets/438773d0-87d7-43e3-a097-f8a115be091c

**Comment by atds55** [comment]
> Você pode fornecer algumas informações de teste para esse mau funcionamento? Preste atenção em ‘remover saroo’

Yes, every time I pressed X on KOF95 without Saroo connected to the Saturn, Fenrir displayed 27 tracks for KOF95.


https://github.com/user-attachments/assets/b34018ba-622c-4a1c-a9ad-8dbfd3219e39


https://github.com/user-attachments/assets/b0457ac2-310f-40c0-8373-4a4fc503b9f6

**Comment by ragingchaos** [comment]
> To use fenrir, you need to press the X key, but this mode is unstable and may not guarantee the accuracy of the data.

This was not a problem in 0.6 and still works when going back to that firm. Whatever changes were recently made in 0.7 are now conflicting with how it loads now. Could it be the "[检测LOADDISC的返回值]" in main.c?

When using Fenrir Loader Kai and pressing X in Fenrir's vanilla menu it boots back into Fenrir Loader Kai, this makes me think it has to be something with the new "Check the return value of LOADDISC" function.

**Comment by tzmwx** [comment]
> > 你能为这个主工作提供一些测试信息吗？“去纱丽”的“重新开始”
> 
> 是的，每次我在没有Saroo连接到土星的情况下在KOF95上按下X，Fenrir都会显示KOF95的27条轨道。
> 
>  73.mp4 
> 
>  16.mp4

The Fenrir you are using should be an old version with a blue PCB. You can use the X key and press A to enter the EXRAM game.

> > Você pode fornecer algumas informações de teste para esse mau funcionamento? Preste atenção em ‘remover saroo’
> 
> Yes, every time I pressed X on KOF95 without Saroo connected to the Saturn, Fenrir displayed 27 tracks for KOF95.
> 
>  73.mp4 
> 
>  16.mp4

The Fenrir you are using should be an old version with a blue PCB. You can use the X key and press A to enter the EXRAM game.

New 20P/21P red PCB FENRIR can use X mode, but there will be no background music in the game because there is a CD image reading error after pressing the X key, and only 2 tracks will appear.

**Comment by whitezombie2000** [comment]
> > Fenrir Loader Kai menu don't seem to play nice together
> 
> A temp work around is to try and load FLK cheats with the X button and when the cheat menu pops up press ABC+Start to return to the CD player menu. Clicking "Start Application" from there loads the game correctly.

Very nice find! I can confirm this workaround works with the old 20 pin Fenrir, Fenrir Loader Kai Menu, and Saroo firmware v.07. I successfully loaded Street Fighter Zero 3 and X-Men vs. Street Fighter with the Saroo.

**Comment by atds55** [comment]
> > > Você pode fornecer algumas informações de teste para esse mau funcionamento? Preste atenção em ‘remover saroo’
> > 
> > 
> > Yes, every time I pressed X on KOF95 without Saroo connected to the Saturn, Fenrir displayed 27 tracks for KOF95.
> > 73.mp4
> > 16.mp4
> 
> The Fenrir you are using should be an old version with a blue PCB. You can use the X key and press A to enter the EXRAM game.
> 
> New 20P/21P red PCB FENRIR can use X mode, but there will be no background music in the game because there is a CD image reading error after pressing the X key, and only 2 tracks will appear.

![image](https://github.com/user-attachments/assets/5d994565-5dd7-44ef-9a0d-596a26422a4a)
My fenrir is the 20pin model (black), it's the same as the image above.

**Comment by whitezombie2000** [comment]
I have the same black pcb 20 pin model. The blue pcb was a 21 pin revision.

21 Pin
![z7oekvexag271](https://github.com/user-attachments/assets/ae7e3463-eb4f-4a0d-8933-d6fb2bcd1746)

21 Pin
![1519039439](https://github.com/user-attachments/assets/38d51610-55f9-4b5e-8d4c-132c3c4c897f)

**Comment by ragingchaos** [comment]
I'm not sure why we are being gaslit with a stability reason the disc detection was working in 0.6. 

Everyone please edit your fenrir.cfg file to:

`	"use_reset":	false,`

and you should be able to load games again with the X button. The A button loader (and Fenrir Loader Kai) are still incompatible.

It seems the recent changes to Saroo has broken this setting so if you need it to stay "true" then stay on firm 0.6 until Saroo's devs can figure out what they need to adjust.

Current State: Fenrir's A, C, and Z button loaders are incompatible along with Fenrir Loader Kai's JHL loader on RAM games and the setting "use_reset:  false" must be set in fenrir.cfg.

**Comment by TrekkiesUnite118** [comment]
> I'm not sure why we are being gaslit with a stability reason when the disc detection was working without issue in 0.6.

Yeah, if the only variable that's changed here is Saroo's firmware, then it's safe to say that Saroo is where the problem lies. If they don't want to do anything on their end to fix it that's fine, but then they need to be honest and say they don't support Saroo working with ODEs like Fenrir.

**Comment by tpunix** [comment]
From firm 0.6 to 0.7, we support load CD-R disc. A lot of code be rewrite for this. It's a hard work. Maybe these changes make other ODEs "not happy". But support other ODEs is not the main purpose. We'll reseaech this when we have more time.

**Comment by TrekkiesUnite118** [comment]
So looking at the code I may have found the issue. It appears that even when booting from a CD Saroo is applying it's game specific patches. To test this theory I had someone burn a modified image of Vampire Savior to a disc. The image changed the Product ID in IP.BIN to a value that doesn't match anything in game_patch.c. Upon testing this the game booted with the same kind of instabilities and graphical corruptions as are being reported here:

https://www.youtube.com/watch?v=FERV8aXIm6k

If I had to take a guess, what's probably going on here is that Fenrir is not able to apply these patches on it's own, but the CD player screen can since as far as I can tell it looks like some of the call stack pointers are modified by Saroo to have it apply these patches when booting a CD from the CD Player screen. 

So to fix this there's probably two options that can be taken:

1. Find some way to get Fenrir to apply these game specific patches so they can boot. 
2. Fix the RAM cart functionality so it accurately behaves the same way as the official RAM carts. This would allow you to eliminate almost all the game specific patches that deal with the RAM and ROM cart games, as they all appear to be doing the same thing. That being changing the A-BUS register settings for interacting with the RAM cart. This would allow Saroo to passively behave as a RAM cart and work properly with both CDs and other ODEs without any need for patching these specific games.

**Comment by ragingchaos** [comment]
To follow up on this, Fenrir users currently have to load games via chainloading.

For Normal Fenrir Menu:

Fenrir.cfg: Set use_reset to false
Saroo: Select 'Load Game Disc' to boot Fenrir's Menu.
Fenrir: Press the Z button to load the game.
Saroo: Select 'Load Game Disc' to load the game you selected with Fenrir.

Fenrir Loader Kai:

Saroo: Select 'Load Game Disc' to boot Fenrir Loader Kai.
Fenrir Kai: Press the Z button to load the game.
'Select Loader' screen: Use the reset command (ABC+Start, Start) or press the Saturn's Reset button to return to Saroo's menu so it can apply its RAM patch.
Saroo: Use 'Load Game Disc' to load the game you selected with Fenrir Loader Kai.
